### PR TITLE
[TS-4160] Reset the txn request/response handles at each hook, since …

### PR DESCRIPTION
**Pull request by sudheerv**
…core may destroy them.

@bgaff @myraid @boaz-r : please review

### CODE REVIEW

**Comment by myraid** [comment]
:+1:  Need to make sure this goes along with #442

**Comment by sudheerv** [comment]
@bgaff can you pls merge this and #442 together, once you review.

File: lib/atscppapi/src/utils_internal.cc [hunk 1]
```diff
@@ -61,19 +73,14 @@ handleTransactionEvents(TSCont cont, TSEvent event, void *edata)
...
-    utils::internal::initTransactionServerResponse(transaction);
-    break;
   case TS_EVENT_HTTP_SEND_RESPONSE_HDR:
-    utils::internal::initTransactionClientResponse(transaction);
-    break;
   case TS_EVENT_HTTP_READ_CACHE_HDR:
-    utils::internal::initTransactionCachedRequest(transaction);
-    utils::internal::initTransactionCachedResponse(transaction);
+    // the buffer handles may be destroyed in the core during redirect follow
+    initTransactionHandles(transaction);
```
Comment (bgaff) [inline lib/atscppapi/src/utils_internal.cc]: I'm not seeing why you want to do them all upfront and repeat in each hook?

**Comment by sudheerv** [comment]
Bryan, the problem is that a lot of these handles are destroyed/recreated at various points in the core. Resetting them at each possible hook allows to have the latest handles in the cppapi objects. Storing them through the life of Transaction without updating them regularly will lead to holding older (and possibly reassigned) handles which could cause corrupting other unrelated  transaction's data/heap.

**Comment by bgaff** [comment]
@sudheerv got it. Might it make sense then just to remove the txn handle caching and just look it up everywhere it's needed? But otherwise i'm :+1: on this.

**Comment by zwoop** [comment]
I'm having a major deja vu here :). We had this discussion at a summit, at LinkedIn, didn't we? :) Yeah, you have to be careful caching these, for these exact reasons. Once a cont is reenabled, things can change under the hood.

**Comment by sudheerv** [comment]
Agree - I don't particularly like caching these either, given that there's no guarantee offered (by the TS API) on the scope of these (the C plugins typically acquire/release these handles within the same scope/hook). 

Removing them altogether probably requires more effort, especially, if there are existing plugins that rely on them and can cause compatibility issues. I'll open a new jira (for 7.0?) to remove these completely, but, in the mean time, re-initing at each hook seems reasonable to avoid holding to stale references?

**Comment by bgaff** [comment]
Why would this need to go in 7.0? it'd be a fully backwards compatible change and could really land at any point.

**Comment by sudheerv** [comment]
Does it not affect the plugins that use API such as Request::getUrl() or Request::getHeaders() etc which rely on the cached handles at the moment? OR Are you suggesting these API could be modified too without necessarily breaking compatibility?

**Comment by bgaff** [comment]
If the handles change then they wouldn't be valid to use anyway so I'm not sure that we need to worry about that, right?

**Comment by sudheerv** [comment]
Yeah, that was the case before this PR, but, the handles are now reinitialized at each hook (along with the URL and header objects), no?

**Comment by bgaff** [comment]
What I'm suggesting is actually pulling the handles when you do something that accesses them...but I suppose what this PR is doing is probably more efficient

**Comment by sudheerv** [comment]
Yeah, I thought about that, but, once you pull the handles you'd need to keep track of it, and reset/reinit them subsequently anyway. So, I tried to keep it simpler by always doing it, at each hook. I do agree though, that we should get rid of caching these handles altogether, just wasn't 100% sure, if it'd affect any existing plugins. Will open a jira and let Boaz research on if compatibility can still be ensured and handles removed safely. Does that sound reasonable?

**Comment by bgaff** [comment]
Sounds good :+1:

**Pull request closed without merging**